Add WeakIValue, use in tracer.

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #21675 Make script::Method a value type
* #21674 Remove lowered execution
* #21565 make tests pass with enable_first_class_module() enabled.
* #21660 Update quantization to work with first-class modules.
* **#21515 Add WeakIValue, use in tracer.**

In order to insert first-class calls to methods, we need to keep
ModuleObject -> Value* mappings in the tracer. To do this, this PR
creates a WeakIValue to replace WeakTensor, and changes the tracer
to use a WeakIValue -> Value* map.
wq

Differential Revision: [D15719981](https://our.internmc.facebook.com/intern/diff/D15719981)

Commit:
Add WeakIValue, use in tracer.

In order to insert first-class calls to methods, we need to keep
ModuleObject -> Value* mappings in the tracer. To do this, this PR
creates a WeakIValue to replace WeakTensor, and changes the tracer
to use a WeakIValue -> Value* map.
wq

Commit:
Update on "Add WeakIValue, use in tracer."

Add WeakIValue, use in tracer.

In order to insert first-class calls to methods, we need to keep
ModuleObject -> Value* mappings in the tracer. To do this, this PR
creates a WeakIValue to replace WeakTensor, and changes the tracer
to use a WeakIValue -> Value* map.
wq

gh-metadata: pytorch pytorch 21515 gh/zdevito/51/head

Commit:
Update on "Add WeakIValue, use in tracer."

Add WeakIValue, use in tracer.

In order to insert first-class calls to methods, we need to keep
ModuleObject -> Value* mappings in the tracer. To do this, this PR
creates a WeakIValue to replace WeakTensor, and changes the tracer
to use a WeakIValue -> Value* map.
wq

gh-metadata: pytorch pytorch 21515 gh/zdevito/51/head

Commit:
Update on "Add WeakIValue, use in tracer."

Add WeakIValue, use in tracer.

In order to insert first-class calls to methods, we need to keep
ModuleObject -> Value* mappings in the tracer. To do this, this PR
creates a WeakIValue to replace WeakTensor, and changes the tracer
to use a WeakIValue -> Value* map.
wq

gh-metadata: pytorch pytorch 21515 gh/zdevito/51/head

zdevito committed:
Update on "Add WeakIValue, use in tracer."

Add WeakIValue, use in tracer.

In order to insert first-class calls to methods, we need to keep
ModuleObject -> Value* mappings in the tracer. To do this, this PR
creates a WeakIValue to replace WeakTensor, and changes the tracer
to use a WeakIValue -> Value* map.
wq

gh-metadata: pytorch pytorch 21515 gh/zdevito/51/head

zdevito committed:
Update on "Add WeakIValue, use in tracer."

Add WeakIValue, use in tracer.

In order to insert first-class calls to methods, we need to keep
ModuleObject -> Value* mappings in the tracer. To do this, this PR
creates a WeakIValue to replace WeakTensor, and changes the tracer
to use a WeakIValue -> Value* map.
wq

gh-metadata: pytorch pytorch 21515 gh/zdevito/51/head

zdevito committed:
Update on "Add WeakIValue, use in tracer."

Add WeakIValue, use in tracer.

In order to insert first-class calls to methods, we need to keep
ModuleObject -> Value* mappings in the tracer. To do this, this PR
creates a WeakIValue to replace WeakTensor, and changes the tracer
to use a WeakIValue -> Value* map.
wq

gh-metadata: pytorch pytorch 21515 gh/zdevito/51/head

zdevito committed:
Update on "Add WeakIValue, use in tracer."

Add WeakIValue, use in tracer.

In order to insert first-class calls to methods, we need to keep
ModuleObject -> Value* mappings in the tracer. To do this, this PR
creates a WeakIValue to replace WeakTensor, and changes the tracer
to use a WeakIValue -> Value* map.

gh-metadata: pytorch pytorch 21515 gh/zdevito/51/head

zdevito committed:
Update on "Add WeakIValue, use in tracer."

Add WeakIValue, use in tracer.

In order to insert first-class calls to methods, we need to keep
ModuleObject -> Value* mappings in the tracer. To do this, this PR
creates a WeakIValue to replace WeakTensor, and changes the tracer
to use a WeakIValue -> Value* map.

gh-metadata: pytorch pytorch 21515 gh/zdevito/51/head

zdevito committed:
Update on "Add WeakIValue, use in tracer."

Add WeakIValue, use in tracer.

In order to insert first-class calls to methods, we need to keep
ModuleObject -> Value* mappings in the tracer. To do this, this PR
creates a WeakIValue to replace WeakTensor, and changes the tracer
to use a WeakIValue -> Value* map.

gh-metadata: pytorch pytorch 21515 gh/zdevito/51/head

zdevito commented on c10/util/intrusive_ptr.h:
```diff
@@ -720,10 +726,7 @@ namespace intrusive_ptr {
 namespace weak_intrusive_ptr {
 
   inline void incref(weak_intrusive_ptr_target* self) {
-    auto wptr = c10::weak_intrusive_ptr<weak_intrusive_ptr_target>::reclaim(self);
```
This change is necessary because ::reclaim would fail if intrusive_ptr_target has never been a weak ptr before. WeakIValue calls incref in the situation we are creating a weak_ptr from a strong pointer when none has existing before. This ensures the method do what it says in its name.

zdevito committed:
Update on "Add WeakIValue, use in tracer."

Add WeakIValue, use in tracer.

In order to insert first-class calls to methods, we need to keep
ModuleObject -> Value* mappings in the tracer. To do this, this PR
creates a WeakIValue to replace WeakTensor, and changes the tracer
to use a WeakIValue -> Value* map.

gh-metadata: pytorch pytorch 21515 gh/zdevito/51/head

suo commented on aten/src/ATen/core/ivalue.h:
```diff
@@ -399,9 +395,106 @@ struct CAFFE2_API IValue final {
+    IValue::Payload pl;
+    pl.as_intrusive_ptr = temp.lock().release();
+    temp.release();
+    if (!pl.as_intrusive_ptr) {
```
Can we do this check before doing the ownership surgery?

suo reviewed (approved):
looks straightforward to me

zdevito committed:
Update on "Add WeakIValue, use in tracer."

Add WeakIValue, use in tracer.

In order to insert first-class calls to methods, we need to keep
ModuleObject -> Value* mappings in the tracer. To do this, this PR
creates a WeakIValue to replace WeakTensor, and changes the tracer
to use a WeakIValue -> Value* map.

gh-metadata: pytorch pytorch 21515 gh/zdevito/51/head
